### DISCUSSION TITLE
Discussion: Methods of preventing elements from loading

### DISCUSSION THREAD

**Discussion by Perseus333** [Pedro-Gregorio/TubeMod#62] [Ideas]
Hi! so this is going to be a discussion for the PR #61 and how to implement it. I figured this should be better because it might be a long conversation and it might be more organized here.

First of all, it's possible! I got it working on Firefox with hard-coded values to delete, and it successfully blocked elements from loading.

### How it (should) work

From what I've found out, it seems like YouTube loads their elements from this file: `https://www.youtube.com/s/desktop/c4395877/jsbin/desktop_polymer.vflset/desktop_polymer.js`. From what I've seen it uses templates to generate the content. The issue with this is that it is a single file and not separate atomic files that we can block with just webRequest. Luckily, it is possible, but I haven't figured it out yet.

So the steps are:

1. It has an event listener on `onBeforeRequest` that looks for this document and blocks it
2. If it succeeds it copies the text from the document and modifies it with regex
3. Then it tries to append a script to the website with the contents of the modified previous document

### Issues

- For now, I haven't been able to properly execute step 2-3 but I think I'm close.

- This just works on Firefox (for now), because it still has compatibility of webRequest but on chrome(ium), declarativeNetRequest is needed. Now, this works also (in theory, haven't tested) on Firefox, as of recently, so maybe we should use that: However, I don't know if this would cause any compatibility issues with browsers relying on older versions of their parent forks.

- That for each element that we have decided to block with css using xpath, now needs to have a new path for deletion using regex to yank it out of the "generator" document.

### Conclusion

I think it might work but it will be a lot of headaches. Also, during my testing it managed to crash my browser (never happened to me before), but I'm *sure* it had nothing to do with the extension. Finally, I will attach the code here to review it and improve on it!

**Comment by Perseus333** [comment]
# Code changes

### background.js

Added this to the end of the file

```javascript
// The "generator" document
const URLToIntercept = "https://www.youtube.com/s/desktop/c4395877/jsbin/desktop_polymer.vflset/desktop_polymer.js";

chrome.webRequest.onBeforeRequest.addListener(
  function(details) {
    if (details.url === URLToIntercept) {
      return { cancel: true };
    }
  },
  { urls: ["*://www.youtube.com/*", "*://m.youtube.com/*"] },
  ["blocking"]
);

// Maybe this should be integrated into the first function?
chrome.webRequest.onCompleted.addListener(
  function(details) {
    if (details.url === URLToIntercept) {
      fetch(details.url)
        .then(response => response.text())
        .then(content => {
          // Arbitrary Hard-Coded Selection (Should delete the suggestions)
          let modifiedJS = content.replace(/da\.Object\.defineProperties\(Z\$\,\{template:\{.*?\}\);\s*/g, "");
          chrome.storage.local.set({ modifiedContent: modifiedJS }, function() {
            console.log("Content modified and saved");
            chrome.tabs.query({ active: true, currentWindow: true }, function(tabs) {
              chrome.tabs.sendMessage(tabs[0].id, { modifiedContent: modifiedJS });
            });
          });
        });
    }
  },
  { urls: ["*://www.youtube.com/*", "*://m.youtube.com/*"] }
);

// Trying to insert the script (not so sure about this, specially the timing)
chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
  if (request.action === "injectModifiedContent") {
    const script = document.createElement('script');
    script.textContent = request.modifiedContent;
    document.head.appendChild(script);
    sendResponse({ status: "success" });
    console.log("Created the script!")
  }
});
``` 

### manifest.json

Added the `webRequest` permisions

```json
  "permissions": ["tabs", "activeTab", "storage", "webRequest", "webRequestBlocking"],
``` 

### Content.js

Added at the end:

```javascript
chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
  if (request.modifiedContent) {
    const script = document.createElement('script');
    script.textContent = request.modifiedContent;
    document.head.appendChild(script);
    sendResponse({ status: "success" });
  }
});
```

**Comment by Perseus333** [comment]
Today I have realized that this is not possible in manifest V3 so, this feature won't be able to be implemented on Chrome. On Firefox however, we could set the manifest to V2, but I don't know if that's the best practice. Either way, I think this feature is too complex for what it does. Not for the webRequest fault but for the way that the content is loaded. It is a 30k+ lines of code, millions of characters and to search in there for elements is a nightmare on its own. 

For these reasons I don't think it's worth it to try to implement them into the extension, at least for the forseable future. But @Pedro-Gregorio let me know what you think.

**Reply by Pedro-Gregorio** [reply]
Yep, looks like we're postponing it. The focus should remain hiding content, at least for now :)
